### PR TITLE
Add script for parsing old articles to populate learncpan CSV file

### PR DESCRIPTION
**Pull request by kjt299**

### CODE REVIEW

**Comment by thibaultduponchelle** [comment]
Great!

happy-santa

I would have probably named it differently (e.g. `advent2learncpan.pl`, `feed-learncpan.pl` or simply `learncpan.pl`)

I tested locally on my side, it's working well:
```
$ perl script/article_parser.pl 
Not a module: get, head, getprint, getstore, mirror at https://perladvent.org/2000/2/ 
Not a module: Template Toolkit at https://perladvent.org/2000/24/ 
Module not found at https://perladvent.org/2004/22nd/ 
Module not found at https://perladvent.org/2004/23rd/ 
Module not found at https://perladvent.org/2004/24th/ 
Module not found at https://perladvent.org/2004/25th/ 
Couldn't process https://perladvent.org/2007/10/ 
Couldn't process https://perladvent.org/2007/12/ 
Couldn't process https://perladvent.org/2007/13/ 
Couldn't process https://perladvent.org/2007/14/ 
Couldn't process https://perladvent.org/2007/15/ 
Couldn't process https://perladvent.org/2007/20/ 
Couldn't process https://perladvent.org/2007/21/ 
Couldn't process https://perladvent.org/2007/22/ 
Couldn't process https://perladvent.org/2008/7/ 
Couldn't process https://perladvent.org/2008/9/ 
Couldn't process https://perladvent.org/2008/17/ 
Couldn't process https://perladvent.org/2009/5/ 
Couldn't process https://perladvent.org/2009/7/ 
Couldn't process https://perladvent.org/2009/9/ 
Couldn't process https://perladvent.org/2009/10/ 
Couldn't process https://perladvent.org/2009/11/ 
Couldn't process https://perladvent.org/2009/12/ 
Couldn't process https://perladvent.org/2009/13/ 
Couldn't process https://perladvent.org/2009/14/ 
Couldn't process https://perladvent.org/2009/15/ 
Couldn't process https://perladvent.org/2009/18/ 
Couldn't process https://perladvent.org/2009/20/ 
Couldn't process https://perladvent.org/2009/21/ 
Couldn't process https://perladvent.org/2009/22/ 
Couldn't process https://perladvent.org/2009/23/ 
Couldn't process https://perladvent.org/2010/25/ 
Not a module: the -M switch at https://perladvent.org/2011/2011-12-05.html 
Not a module: List-Util & Friends at https://perladvent.org/2011/2011-12-10.html 
Not a module: App-Uni & Unicode-Tussle at https://perladvent.org/2011/2011-12-24.html 
Not a module: Perl Advent 2011 at https://perladvent.org/2011/2011-12-25.html 
Not a module: perl -pi -e at https://perladvent.org/2012/2012-12-02.html 
Not a module: fatpack / App-FatPacker at https://perladvent.org/2012/2012-12-14.html 
Not a module: forkprove and Test-Pretty at https://perladvent.org/2012/2012-12-20.html 
Not a module: Perl Advent 2012 at https://perladvent.org/2012/2012-12-25.html 
Not a module: RDF-Trine and RDF-Query at https://perladvent.org/2013/2013-12-01.html 
Not a module: Dancer's Megasplat at https://perladvent.org/2013/2013-12-02.html 
Not a module: PPI-HTML and Text-VimColor at https://perladvent.org/2013/2013-12-20.html 
Not a module: hub, bash at https://perladvent.org/2013/2013-12-21.html 
Not a module: Perl Advent 2013 at https://perladvent.org/2013/2013-12-25.html 
Not a module: Unicode CLDR  at https://perladvent.org/2014/2014-12-23.html 
Not a module: Perl Advent 2014 at https://perladvent.org/2014/2014-12-25.html 
Not a module: Perl Advent 2015 at https://perladvent.org/2015/2015-12-25.html 
Not a module: Linux-Clone Linux-Unshare Linux-Setns at https://perladvent.org/2016/2016-12-19.html 
Not a module: Test2 Tooling at https://perladvent.org/2016/2016-12-24.html 
Not a module: Perl Advent 2016 at https://perladvent.org/2016/2016-12-25.html 
Not a module: Printing Unicode with Perl at https://perladvent.org/2017/2017-12-01.html 
Not a module: Enforcing coding standards via Perl-Critic at https://perladvent.org/2017/2017-12-03.html 
Not a module: Modern Heredocs at https://perladvent.org/2017/2017-12-04.html 
Not a module: The Babycart Operator at https://perladvent.org/2017/2017-12-05.html 
Not a module: Recursive Regular Expressions with Named Groups at https://perladvent.org/2017/2017-12-09.html 
Not a module: MooseX-LazyRequire, MooseX-UndefTolerant-Attribute at https://perladvent.org/2017/2017-12-19.html 
Not a module: Custom Relationships with DBIx-Class at https://perladvent.org/2017/2017-12-22.html 
Not a module: Merry Christmas! at https://perladvent.org/2017/2017-12-25.html 
Not a module: SVG (and SVG-ChristmasTree) at https://perladvent.org/2018/2018-12-12.html 
Not a module: Future-based interfaces at https://perladvent.org/2018/2018-12-14.html 
Not a module: App-HTTPThis and App-HTTPSThis at https://perladvent.org/2018/2018-12-24.html 
Not a module: Merry Christmas at https://perladvent.org/2018/2018-12-25.html 
Not a module: FFI-Platypus and Go at https://perladvent.org/2019/2019-12-04.html 
Not a module: Phillips Hue Home Bridge at https://perladvent.org/2019/2019-12-05.html 
Not a module: iTerm and Perl at https://perladvent.org/2019/2019-12-20.html 
Not a module: It's Chrismas at https://perladvent.org/2019/2019-12-25.html 
Not a module: SVG (and SVG-ChristmasTree) at https://perladvent.org/2020/2020-12-13.html 
Not a module: Christmas 2020 at https://perladvent.org/2020/2020-12-25.html 
```

## `perladvent.org` needs a sitemap!

**Comment by kjt299** [comment]
Relates to #141

File: script/learncpan.pl [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+        my $url = get_url( $year, $day );
+        get_articles( $url,
+            { title_selector => 'h1.title', module_selector => 'div.subtitle' }
+        );
+    }
+}
+
+my @all    = csv( in => "../learncpan/tutorials.csv" );
+my @merged = ( @{ $all[0] }, @rows );
```
Comment (thibaultduponchelle) [inline script/learncpan.pl:91]: What is doing exactly this line?

File: script/learncpan.pl [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    unless ($module) {
+        print "Module not found at $url \n";
+        return;
+    }
+
+    $title  =~ s/,//g;
+    $module =~ s/\s-\s.+//;
+    $module =~ s/::/-/g;
+
+    #If $module stil has white space at this point it probably isn't a module
```
Comment (thibaultduponchelle) [inline script/learncpan.pl:118]: I feel better with space after #

**Review by thibaultduponchelle — APPROVED** [review]
I carefully read this script, I approve 👍🏼

File: script/learncpan.pl [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        my $url = get_url( $year, $day );
+        my $dom = get_dom($url);
+        next unless $dom;
+        my $title = $dom->at('title');
+
+        push @rows, [ 'TODO', $url, $title ];
+    }
+}
+
+foreach my $year ( 2011 .. 2020 ) {
```
Comment (oalders) [inline script/learncpan.pl:80]: I think we can skip 2020 as it was (if I'm correct) a collection of highlights from previous years.

File: script/learncpan.pl [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    my $module = $ms->text if $ms;
+
+    unless ($module) {
+        print "Module not found at $url \n";
+        return;
+    }
+
+    $title  =~ s/,//g;
+    $module =~ s/\s-\s.+//;
+    $module =~ s/::/-/g;
```
Comment (oalders) [inline script/learncpan.pl:116]: Some of the modules that are being skipped are articles about multiple modules. For instance

> Not a module: MooseX-LazyRequire, MooseX-UndefTolerant-Attribute

We could handle *some* of these via something like:

```perl
my (@mods) = $module =~ m{\w+\-\w+}g;
```

and then adding a row for each module.

**Review by oalders — CHANGES_REQUESTED** [review]
Thanks so much for your submission. 🐫 

I had a couple of suggestions.

File: script/learncpan.pl (hunk 4 above)
Comment (thibaultduponchelle) [reply script/learncpan.pl:116]: ++

File: script/learncpan.pl (hunk 3 above)
Comment (thibaultduponchelle) [reply script/learncpan.pl:80]: +1

File: script/learncpan.pl (hunk 1 above)
Comment (kjt299) [reply script/learncpan.pl:91]: It is meant to merge existing tutorials with the ones we want to add into one array so we can easily create a new csv file.

File: script/learncpan.pl (hunk 4 above)
Comment (kjt299) [reply script/learncpan.pl:115]: Makes sense :)

**Review by thibaultduponchelle — CHANGES_REQUESTED** [review]
As per #141 I think year 2000 and 2001 should be ignored.

**Comment by kjt299** [comment]
Removed 2000 and 2001 as suggested :)

**Review by thibaultduponchelle — APPROVED** [review]

**Comment by thibaultduponchelle** [comment]
Just one little last thing and I will merge it (except if @oalders is not OK)

The ->text removes the `<title>....</title>` from TODO lines in CSV

**Comment by oalders** [comment]
@thibaultduponchelle please merge once you are happy. Thanks again @kjt299. ⭐

**Comment by thibaultduponchelle** [comment]
@oalders Can you please "approve" this PR so I can merge it.

**Review by oalders — APPROVED** [review]

**Pull request merged**